cpp/arcticdb/column_store/memory_segment_impl.cpp

```diff
@@ -428,24 +431,41 @@ std::shared_ptr<SegmentInMemoryImpl> SegmentInMemoryImpl::truncate(size_t start_
     auto output = std::make_shared<SegmentInMemoryImpl>();
 
     output->set_row_data(num_values - 1);
-    output->set_string_pool(string_pool_);
     output->set_compacted(compacted_);
+    if (!reconstruct_string_pool) {
+        output->set_string_pool(string_pool_);
+    }
     if (metadata_) {
         google::protobuf::Any metadata;
         metadata.CopyFrom(*metadata_);
         output->set_metadata(std::move(metadata));
     }
 
-    for(const auto&& [idx, column] : folly::enumerate(columns_)) {
-        auto truncated_column = Column::truncate(column, start_row, end_row);
-        output->add_column(descriptor_->field(idx), truncated_column);
+    for (const auto&& [idx, column] : folly::enumerate(columns_)) {
+        const DataType column_type = column->type().data_type();
+        const Field& field = descriptor_->field(idx);
+        std::shared_ptr<Column> truncated_column = Column::truncate(column, start_row, end_row);
+
+        if (is_sequence_type(column_type) && reconstruct_string_pool) {
+            ChunkedBuffer& truncated_buffer = truncated_column->data().buffer();
+            for (position_t row = 0; row < truncated_column->row_count(); ++row) {
```

Comment:
`get_offset_string_at` will be doing a random-access into the chunked buffer, which adds a bunch of unnecessary branching when we are just iterating forwards like this. You can use the new `Column::for_each` [method](https://github.com/man-group/ArcticDB/blob/master/cpp/arcticdb/column_store/column.hpp#L661-L668) to make this both more efficient and more readable.